### PR TITLE
Fix: ES returns too many headers during bundle deletion (#2455)

### PR DESCRIPTION
**Pull request by amarjandu**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [ ] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [ ] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Added PR reference to merge commit title
- [ ] Collected commit title tags in merge commit title
- [ ] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: src/azul/indexer/document.py
Comment (amarjandu) [inline src/azul/indexer/document.py:464]: Might want to change this to be `None` rather than (None, None).

**Review by danielsotirhos — APPROVED** [review]
Surgically done, looks clean.

**Comment by amarjandu** [comment]
@danielsotirhos just some modifications to how the version is figured out during the `from_index` operation. There should be two cases that can exist, either the `seq_no` and `primary_term` are not set (missing from the hits), or they are both set, one cannot be set without the other.

**Review by danielsotirhos — APPROVED** [review]

**Comment by amarjandu** [comment]
@hannes-ucsc please hold off on the review, i need to push the drop! commit to see the failure in travis. The rebase removed the previous failure notification.

File: src/azul/indexer/document.py
Comment (hannes-ucsc) [inline src/azul/indexer/document.py:349]:
```suggestion
    # For VersionType.internal, version is a tuple composed of the sequence number and primary term. For VersionType.none and .create_only, it is None.
```

But check the correctness of this statement.

File: src/azul/indexer/document.py
Comment (hannes-ucsc) [inline src/azul/indexer/document.py:350]: Are you sure both are integers?

File: src/azul/indexer/document.py
Comment (hannes-ucsc) [inline src/azul/indexer/document.py:528]: See, here `self.version_type` is `VersionType.none` but you're extracting a version. That's very counter intuitive and doesn't quite add up.

File: src/azul/indexer/document.py
Comment (hannes-ucsc) [inline src/azul/indexer/document.py:550]:
```suggestion
                # For internal versioning, self.version is None for new documents
                result['if_seq_no'], result['if_primary_term'] = self.version
```

File: test/docker_container_test_case.py
Comment (hannes-ucsc) [inline test/docker_container_test_case.py:107]: What would such a line of the container output be, for example?

Comment (hannes-ucsc) [reply test/docker_container_test_case.py:107]: And isn't this useful in general? If it is, this change shouldn't be dropped.

File: src/azul/indexer/index_service.py
Comment (hannes-ucsc) [inline src/azul/indexer/index_service.py:353]: Still brittle coupling, Assumes that only contributions use internal versioning. If for some reason we switched aggregates to use internal versioning, too, things would break.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/docker_container_test_case.py
Comment (amarjandu) [reply test/docker_container_test_case.py:107]: The line within the container logs looks like:
`'[2021-01-28T01:49:24,844][WARN ][o.e.d.c.ParseField       ] [9Bhi2zf] Deprecated field [exclude] used, expected [excludes] instead')`

The AssertionError in python is more verbose:
`AssertionError: ["{message : DeprecationWarning('[2021-01-28T01:49:24,844][WARN ][o.e.d.c.ParseField       ] [9Bhi2zf] Deprecated field [exclude] used, expected [excludes] instead'), category : 'DeprecationWarning', filename : '/home/travis/build/DataBiosphere/azul/test/docker_container_test_case.py', lineno : 107, line : None}"]`

Yes this feature of extracting deprecation warnings out of containers that we use is useful; additional deprecation warnings would only occur if we change the versioning of Elasticsearch, or modifications to the requirements, so it would be nice to see these warnings. 

If we update to ES7.0, the Elasticsearch-py module will raise these errors so they would be more visible in the Azul logs. https://elasticsearch-py.readthedocs.io/en/7.10.0/Changelog.html#id25

If we switch over to a long-living Elasticsearch instance for unit-tests (#2653) we should try to retain this feature.

File: src/azul/indexer/document.py
Comment (amarjandu) [reply src/azul/indexer/document.py:350]: I cant find documentation of the type expressed in ElasticSearch 6.x documentation, but here is a reference to the bulk API for 7.9 that states that the response `seq_num` and `primary_term` are integers. https://www.elastic.co/guide/en/elasticsearch/reference/7.9/docs-bulk.html#bulk-api-response-body

All references to the sequence_number and primary_term that I have seen in the 6.x documentation do not mention the typing for these parameters. 

unfortunately `elasticsearch-py` is also not helpful it uses `Optional[any]` 
https://github.com/elastic/elasticsearch-py/search?q=if_seq_no

If you would like me to spend more time looking for explicit documentation, I can spend more time searching.

File: src/azul/indexer/index_service.py
Comment (amarjandu) [reply src/azul/indexer/index_service.py:353]: https://github.com/DataBiosphere/azul/blob/3ee6f03ab1115275ec1dcd04af95cb20ff384148/src/azul/indexer/document.py#L591

Looking into this more.

Comment (amarjandu) [reply src/azul/indexer/index_service.py:353]: When aggregates are first created, the aggregate class is formed by the contribution, and the __init__ explicitly has the parameter `version=None` [passed to it.](https://github.com/DataBiosphere/azul/blob/0770e7c96e9a23b9e874745a2df28b29b64fe2a1/src/azul/indexer/index_service.py#L442)

The next time the old_aggregates are fetched, the relative documents for the tallies are obtained with `mget`
https://github.com/DataBiosphere/azul/blob/0770e7c96e9a23b9e874745a2df28b29b64fe2a1/src/azul/indexer/index_service.py#L295
`mget` returns multiple docs with responses that are similar to [get](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/docs-get.html#get-source-filtering). The document objects have their applicable `seq_num` && `primary_term` within the response. This document is used as the parameter for the aggregate_cls.from_index() method, where it is extracted into the version tuple.

File: test/docker_container_test_case.py
Comment (hannes-ucsc) [reply test/docker_container_test_case.py:107]: Remind me to discuss at parking lot. (Note to self: reorder commits, and add a drop commit so the versioning deprecation can be observed before we switch to the new versioning)

Comment (amarjandu) [reply test/docker_container_test_case.py:107]: From standup discussion: 
1.  Add container changes (see all deprecation warnings)
2. Eliminate "other" deprecation warnings ('excludes'...) (see only version deprecation warning)
3. Add fixes for version deprecation warnings

**Review by hannes-ucsc — COMMENTED** [review]
Looks very good.

No need to push separately, next time. I've seen enough. Just the two fixups please.

Commit titles:

`Remove max_header patch for hca_indexer` should refer to the ticket.

`Emit deprecation warnings from test containers` should be `Propagate deprecation warnings from ES container`

`Update deprecated parameters` => `Fix: _source_include is deprecated in ES 6.x`

`Add fixme for #2758` => `Add FIXME (#2758)`

`Add support for seq_num and primary_term (#2455)` => `Support ES versioning with seq_num and primary_term (#2455)`

File: test/azul_test_case.py
Comment (amarjandu) [reply test/azul_test_case.py]: a mistake, cleaned up now.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
@danielsotirhos, while squashing this, please rename commit `Add FIXME (#2758)` to `Ignore GC warning from ES containers (#2758)`.

**Comment by hannes-ucsc** [comment]
@danielsotirhos, I would like to ensure that this doesn't break remote reindexing. Please reindex the sandbox, and only the sandbox, after building this in the sandbox, and before merging it.

**Comment by hannes-ucsc** [comment]
For demo, run a suitable test with an attached debugger and show the versioning fields in an ES response.

**Comment by danielsotirhos** [comment]
As per requested a reindex for this PR was run only on sandbox and not dev.

**Pull request merged**
